R-package/tests/testthat/test_lgb.Booster.R

```diff
@@ -799,37 +799,170 @@ test_that("all parameters are stored correctly with save_model_to_string()", {
+        , "[cegb_penalty_feature_coupled: ]"
+        , "[path_smooth: 0]"
+        , "[interaction_constraints: ]"
+        , sprintf("[verbosity: %i]", .LGB_VERBOSITY)
+        , "[saved_feature_importance_type: 0]"
+        , "[use_quantized_grad: 0]"
+        , "[num_grad_quant_bins: 4]"
+        , "[quant_train_renew_leaf: 0]"
+        , "[stochastic_rounding: 1]"
+        , "[linear_tree: 0]"
+        , "[max_bin: 255]"
+        , "[max_bin_by_feature: ]"
+        , "[min_data_in_bin: 3]"
+        , "[bin_construct_sample_cnt: 200000]"
+        , "[data_random_seed: 2350]"
+        , "[is_enable_sparse: 1]"
+        , "[enable_bundle: 1]"
+        , "[use_missing: 1]"
+        , "[zero_as_missing: 0]"
+        , "[feature_pre_filter: 1]"
+        , "[pre_partition: 0]"
+        , "[two_round: 0]"
+        , "[header: 0]"
+        , "[label_column: ]"
+        , "[weight_column: ]"
+        , "[group_column: ]"
+        , "[ignore_column: ]"
+        , "[categorical_feature: ]"
+        , "[forcedbins_filename: ]"
+        , "[precise_float_parser: 0]"
+        , "[parser_config_file: ]"
+        , "[objective_seed: 4309]"
+        , "[num_class: 1]"
+        , "[is_unbalance: 0]"
+        , "[scale_pos_weight: 1]"
+        , "[sigmoid: 1]"
+        , "[boost_from_average: 1]"
+        , "[reg_sqrt: 0]"
+        , "[alpha: 0.9]"
+        , "[fair_c: 1]"
+        , "[poisson_max_delta_step: 0.7]"
+        , "[tweedie_variance_power: 1.5]"
+        , "[lambdarank_truncation_level: 30]"
+        , "[lambdarank_norm: 1]"
+        , "[label_gain: ]"
+        , "[lambdarank_position_bias_regularization: 0]"
+        , "[eval_at: ]"
+        , "[multi_error_top_k: 1]"
+        , "[auc_mu_weights: ]"
+        , "[num_machines: 1]"
+        , "[local_listen_port: 12400]"
+        , "[time_out: 120]"
+        , "[machine_list_filename: ]"
+        , "[machines: ]"
+        , "[gpu_platform_id: -1]"
+        , "[gpu_device_id: -1]"
+        , "[gpu_use_dp: 0]"
+        , "[num_gpu: 1]"
+    )
+    all_param_entries <- c(non_default_param_entries, default_param_entries)
 
     # parameters should match what was passed from the R package
-    expect_equal(sum(startsWith(params_in_file, "[metric:")), 1L)
-    expect_equal(sum(params_in_file == "[metric: l2]"), 1L)
-
-    expect_equal(sum(startsWith(params_in_file, "[num_iterations:")), 1L)
-    expect_equal(sum(params_in_file == "[num_iterations: 4]"), 1L)
-
-    expect_equal(sum(startsWith(params_in_file, "[objective:")), 1L)
-    expect_equal(sum(params_in_file == "[objective: regression]"), 1L)
-
-    expect_equal(sum(startsWith(params_in_file, "[verbosity:")), 1L)
-    expect_equal(sum(params_in_file == sprintf("[verbosity: %i]", .LGB_VERBOSITY)), 1L)
+    model_str <- bst$save_model_to_string()
+    params_in_file <- .params_from_model_string(model_str = model_str)
+    for (param_str in all_param_entries) {
+        expect_equal(sum(params_in_file == param_str), 1L, info = param_str)
+    }
```

Comment:
I think this is easier to grasp ([source](https://testthat.r-lib.org/reference/expect_setequal.html)):
> expect_in(x, y) tests every element of x is in y (i.e. x is a subset of y).
```suggestion
    expect_in(all_param_entries, params_in_file)
```

Reply:
Love this!! I didn't know about `expect_in`. I especially like the error message it gives, which is better than what I was trying to accomplish with `info = params_str`

```r
testthat::expect_in(c("a", "b", "c"), c("d")
```

```text
Error: c("a", "b", "c") (`actual`) isn't fully contained within c("d") (`expected`).
* Missing from `expected`: "a", "b", "c"
* Present in `expected`:   "d"
```

Reply:
The Windows R 3.6 tests failed as a result of this:

> -- 1. Error (test_lgb.Booster.R:948:5): all parameters are stored correctly with

>  Error: could not find function "expect_in"

([example build link](https://github.com/microsoft/LightGBM/actions/runs/6149355551/job/16685052422?pr=6077))

Looks like `expect_in()` was only added to `{testthat}` 3 months ago, meaning it would have probably been in `{testthat}` 3.1.10: https://github.com/r-lib/testthat/pull/1817

I suspect the R 3.6 jobs are getting an older version of `{testthat}` than that. I think it'd be better to have this more-informative error message and easier-to-understand test code than to have coverage of this on R 3.6 (which by now is 3.5 years old ([link](https://cran.r-project.org/bin/windows/base/old/))).

I'll add a skip in the test + put a floor on `{testthat}` in the package's `DESCRIPTION`.

Reply:
> I'll add a skip in the test + put a floor on `{testthat}` in the package's `DESCRIPTION`

I decided to try going a different direction with this. I added a little helper function in the tests to use `testthat::expect_in()` if it exists (for the slightly more informative error message) and otherwise do this in a base-R way: https://github.com/microsoft/LightGBM/pull/6077/commits/40adeafa2c87ad29ecec82c2ea87c486ebb6fc67

I think that's preferable to fully skipping the tests.

And I realized a floor on `{testthat}` in the `DESCRIPTION` doesn't make sense, if the root of the problem is "we're getting an older version of `{testthat}` in the R 3.6 CI jobs".

Whenever you have time @jmoralez , could you let me know what you think?

Reply:
Totally agree, thanks! Sorry for the extra work, I thought it was going to be easier to use `expect_in` when I suggested it.

Reply:
No need to be sorry, I appreciate the suggestion and I learned some things while working through it!